pulsar-client-cpp/lib/KeyValueImpl.cc

```diff
@@ -0,0 +1,76 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#include <pulsar/Schema.h>
+#include "SharedBuffer.h"
+#include "KeyValueImpl.h"
+
+using namespace pulsar;
+
+namespace pulsar {
+
+KeyValueImpl::KeyValueImpl(const char *data, int length, KeyValueEncodingType keyValueEncodingType) {
+    if (keyValueEncodingType == KeyValueEncodingType::INLINE) {
+        SharedBuffer buffer = SharedBuffer::wrap(const_cast<char *>(data), length);
+        auto keySize = buffer.readUnsignedInt();
+        if (keySize != INVALID_SIZE) {
+            SharedBuffer keyContent = buffer.slice(0, keySize);
+            key_ = std::string(keyContent.data(), keySize);
+            buffer.consume(keySize);
+        }
+        auto valueSize = buffer.readUnsignedInt();
+        if (valueSize != INVALID_SIZE) {
+            valueBuffer_ = buffer.slice(0, valueSize);
+        }
+    } else {
+        valueBuffer_ = SharedBuffer::wrap(const_cast<char *>(data), length);
+    }
+}
+
+KeyValueImpl::KeyValueImpl(std::string &&key, std::string &&value)
+    : key_(std::move(key)), valueBuffer_(SharedBuffer::take(std::move(value))) {}
+
+SharedBuffer KeyValueImpl::getContent(KeyValueEncodingType keyValueEncodingType) {
+    if (keyValueEncodingType == KeyValueEncodingType::INLINE) {
+        auto keySize = key_.length();
+        auto valueSize = valueBuffer_.readableBytes();
+        auto buffSize = sizeof(keySize) + keySize + sizeof(valueSize) + valueSize;
+        SharedBuffer buffer = SharedBuffer::allocate(buffSize);
+        buffer.writeUnsignedInt(keySize == 0 ? -1 : keySize);
+        buffer.write(key_.c_str(), keySize);
+
+        buffer.writeUnsignedInt(valueSize == 0 ? -1 : valueSize);
+        buffer.write(valueBuffer_.data(), valueSize);
+
+        return buffer;
+    } else {
+        return SharedBuffer::copyFrom(valueBuffer_, valueBuffer_.readableBytes());
```

Comment:
Can we just return `valueBuffer_` here to avoid deep copy? I think in this case `valueBuffer_` will keep constant so it doesn't need a deep copy.

Reply:
Here I want to be consistent with the logic when `type==INLINE`, and also avoid destroying the encapsulation of the class.

Users may be used like this:
```
    KeyValue keyValue((std::string(keyData)), std::string(valueData));
    Message msg = MessageBuilder().setContent(keyValue).setProperty("x", "1").build();
    producer.send(msg);
    // continue to use keyValue do something. 
    // keyValue.getValue(); 
```